File: tests/cpp/test_official_kda.cpp
// 공식 KDA portable oracle의 full/incremental state와 모든 경계를 검증합니다.
#include "k3x/official_kda.hpp"

#include <array>
#include <bit>
#include <cmath>
#include <cstddef>
#include <cstdint>
#include <iostream>
#include <limits>
#include <span>
#include <string_view>
#include <vector>

namespace {

std::uint16_t bf16(float value) {
    auto bits = std::bit_cast<std::uint32_t>(value);
    bits += 0x7fffU + ((bits >> 16U) & 1U);
    return static_cast<std::uint16_t>(bits >> 16U);
}

template <std::size_t Size>
std::array<std::uint16_t, Size> words(const std::array<float, Size>& values) {
    std::array<std::uint16_t, Size> result{};
    for (std::size_t index = 0; index < Size; ++index) {
        result[index] = bf16(values[index]);
    }
    return result;
}

bool close(std::span<const float> left, std::span<const float> right,
           float tolerance = 1.0e-6F) {
    if (left.size() != right.size()) return false;
    for (std::size_t index = 0; index < left.size(); ++index) {
        if (std::abs(left[index] - right[index]) > tolerance) return false;
    }
    return true;
}

bool joined_close(std::span<const float> full,
                  std::span<const float> first,
                  std::span<const float> second,
                  float tolerance = 1.0e-6F) {
    std::vector<float> joined(first.begin(), first.end());
    joined.insert(joined.end(), second.begin(), second.end());
    return close(full, joined, tolerance);
}

void print_vector(std::span<const float> values) {
    std::cout << '[';
    for (std::size_t index = 0; index < values.size(); ++index) {
        if (index) std::cout << ',';
        std::cout << values[index];
    }
    std::cout << ']';
}

void print_words(std::span<const std::uint16_t> values) {
    std::cout << '[';
    for (std::size_t index = 0; index < values.size(); ++index) {
        if (index) std::cout << ',';
        std::cout << values[index];
    }
    std::cout << ']';
}

}  // namespace

int main(int argc, char** argv) {
    const k3x::OfficialKdaConfig config{4, 2, 2, 3, 1.0e-5F, -5.0F};
    const auto identity = words<16>({
        1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1});
    const auto k_proj = words<16>({
        0.5F, 0, 0.25F, 0, 0, 0.75F, 0, -0.25F,
        0.25F, 0, 1, 0, 0, -0.5F, 0, 0.5F});
    const std::array<float, 12> q_conv{
        0.25F, -0.5F, 1, -0.25F, 0.5F, 0.75F,
        0.5F, 0.25F, -0.5F, -0.5F, 0.25F, 1.25F};
    std::array<float, 12> k_conv{};
    std::array<float, 12> v_conv{};
    for (std::size_t index = 0; index < q_conv.size(); ++index) {
        k_conv[index] = q_conv[index] * 0.75F;
        v_conv[index] = q_conv[index] * -0.5F;
    }
    const auto f_a = words<8>({
        0.5F, -0.25F, 0, 0.25F, 0, 0.5F, -0.5F, 0.25F});
    const auto f_b = words<8>({
        1, 0, 0.5F, -0.5F, 0, 1, -0.25F, 0.75F});
    const std::array<float, 2> a_log{0, 0.5F};
    const std::array<float, 4> dt_bias{0.25F, -0.5F, 0.75F, -0.25F};
    const auto b_proj = words<8>({
        0.5F, -0.25F, 0.25F, 0, 0, 0.5F, -0.5F, 0.25F});
    const auto g_proj = words<16>({
        0.5F, 0, 0, -0.25F, 0, 0.5F, 0.25F, 0,
        -0.25F, 0, 0.5F, 0, 0, 0.25F, 0, 0.75F});
    const std::array<float, 2> o_norm{1, 1.5F};
    const k3x::OfficialKdaWeightsView weights{
        {identity, 4, 4}, {k_proj, 4, 4}, {identity, 4, 4},
        q_conv, k_conv, v_conv,
        {f_a, 2, 4}, {f_b, 4, 2}, a_log, dt_bias,
        {b_proj, 2, 4}, {g_proj, 4, 4}, o_norm, {identity, 4, 4}};
    const std::array<float, 8> hidden{
        0.5F, -1, 0.25F, 0.75F, -0.25F, 0.5F, 1, -0.5F};
    const auto zero = k3x::zero_official_kda_state(config);
    const auto original_zero = zero;
    const auto full = k3x::official_kda_cpu(hidden, weights, zero, config);
    const auto first = k3x::official_kda_cpu(
        std::span(hidden).first(4), weights, zero, config);
    if (!full || !first) return 1;
    const auto second = k3x::official_kda_cpu(
        std::span(hidden).last(4), weights, first.value().state, config);
    if (!second) return 2;
    std::vector<float> incremental = first.value().output;
    incremental.insert(incremental.end(), second.value().output.begin(),
                       second.value().output.end());
    if (!close(full.value().output, incremental, 0.0F) ||
        full.value().state.conv_q != second.value().state.conv_q ||
        full.value().state.conv_k != second.value().state.conv_k ||
        full.value().state.conv_v != second.value().state.conv_v ||
        !close(full.value().state.recurrent_v_first,
               second.value().state.recurrent_v_first) ||
        zero.conv_q != original_zero.conv_q ||
        zero.conv_k != original_zero.conv_k ||
        zero.conv_v != original_zero.conv_v ||
        zero.recurrent_v_first != original_zero.recurrent_v_first) return 3;
    const auto& all = full.value().boundaries;
    const auto& one = first.value().boundaries;
    const auto& two = second.value().boundaries;
    if (!joined_close(all.projected_q, one.projected_q, two.projected_q, 0.0F) ||
        !joined_close(all.projected_k, one.projected_k, two.projected_k, 0.0F) ||
        !joined_close(all.projected_v, one.projected_v, two.projected_v, 0.0F) ||
        !joined_close(all.convolved_q, one.convolved_q, two.convolved_q, 0.0F) ||
        !joined_close(all.convolved_k, one.convolved_k, two.convolved_k, 0.0F) ||
        !joined_close(all.convolved_v, one.convolved_v, two.convolved_v, 0.0F) ||
        !joined_close(all.q, one.q, two.q, 0.0F) ||
        !joined_close(all.k, one.k, two.k, 0.0F) ||
        !joined_close(all.v, one.v, two.v, 0.0F) ||
        !joined_close(all.log_decay, one.log_decay, two.log_decay) ||
        !joined_close(all.beta, one.beta, two.beta) ||
        !joined_close(all.recurrent_output, one.recurrent_output,
                      two.recurrent_output) ||
        !joined_close(all.gated, one.gated, two.gated, 0.0F)) return 7;

    if (argc == 2 && std::string_view(argv[1]) == "--dump") {
        const auto& result = full.value();
        std::cout.precision(9);
        std::cout << "{\"projected_q\":"; print_vector(result.boundaries.projected_q);
        std::cout << ",\"projected_k\":"; print_vector(result.boundaries.projected_k);
        std::cout << ",\"projected_v\":"; print_vector(result.boundaries.projected_v);
        std::cout << ",\"convolved_q\":"; print_vector(result.boundaries.convolved_q);
        std::cout << ",\"convolved_k\":"; print_vector(result.boundaries.convolved_k);
        std::cout << ",\"convolved_v\":"; print_vector(result.boundaries.convolved_v);
        std::cout << ",\"q\":"; print_vector(result.boundaries.q);
        std::cout << ",\"k\":"; print_vector(result.boundaries.k);
        std::cout << ",\"v\":"; print_vector(result.boundaries.v);
        std::cout << ",\"log_decay\":"; print_vector(result.boundaries.log_decay);
        std::cout << ",\"beta\":"; print_vector(result.boundaries.beta);
        std::cout << ",\"recurrent_output\":"; print_vector(result.boundaries.recurrent_output);
        std::cout << ",\"gated\":"; print_vector(result.boundaries.gated);
        std::cout << ",\"output\":"; print_vector(result.output);
        std::cout << ",\"conv_q\":"; print_words(result.state.conv_q);
        std::cout << ",\"conv_k\":"; print_words(result.state.conv_k);
        std::cout << ",\"conv_v\":"; print_words(result.state.conv_v);
        std::cout << ",\"recurrent_v_first\":";
        print_vector(result.state.recurrent_v_first);
        std::cout << "}\n";
        return 0;
    }

    auto bad_weights = weights;
    bad_weights.a_log = std::span(a_log).first(1);
    if (k3x::official_kda_cpu(hidden, bad_weights, zero, config)) return 8;
    bad_weights = weights;
    const std::array<float, 2> bad_a_log{0, std::numeric_limits<float>::infinity()};
    bad_weights.a_log = bad_a_log;
    if (k3x::official_kda_cpu(hidden, bad_weights, zero, config)) return 9;
    auto bad_state = zero;
    bad_state.recurrent_v_first.push_back(0);
    if (k3x::official_kda_cpu(hidden, weights, bad_state, config)) return 10;
    return 0;
}

File: tests/python/test_official_kda.py
# 공식 KDA scalar oracle의 full/incremental 상태와 channel-wise recurrence를 검증합니다.
from __future__ import annotations

from dataclasses import replace

import pytest
import torch

from k3x_ref.official_kda import (
    OfficialKdaConfig,
    OfficialKdaState,
    OfficialKdaWeights,
    official_kda,
    zero_official_kda_state,
)
from k3x_ref.official_kda import _project


def _config() -> OfficialKdaConfig:
    return OfficialKdaConfig(
        hidden_size=4,
        heads=2,
        head_dim=2,
        conv_width=3,
        rms_norm_epsilon=1.0e-5,
        gate_lower_bound=-5.0,
    )


def _weights() -> OfficialKdaWeights:
    identity = torch.eye(4, dtype=torch.bfloat16)
    conv = torch.tensor(
        [
            [0.25, -0.5, 1.0],
            [-0.25, 0.5, 0.75],
            [0.5, 0.25, -0.5],
            [-0.5, 0.25, 1.25],
        ],
        dtype=torch.float32,
    )
    return OfficialKdaWeights(
        q_proj=identity,
        k_proj=torch.tensor(
            [
                [0.5, 0.0, 0.25, 0.0],
                [0.0, 0.75, 0.0, -0.25],
                [0.25, 0.0, 1.0, 0.0],
                [0.0, -0.5, 0.0, 0.5],
            ],
            dtype=torch.bfloat16,
        ),
        v_proj=identity,
        q_conv=conv,
        k_conv=conv * 0.75,
        v_conv=conv * -0.5,
        f_a_proj=torch.tensor(
            [[0.5, -0.25, 0.0, 0.25], [0.0, 0.5, -0.5, 0.25]],
            dtype=torch.bfloat16,
        ),
        f_b_proj=torch.tensor(
            [[1.0, 0.0], [0.5, -0.5], [0.0, 1.0], [-0.25, 0.75]],
            dtype=torch.bfloat16,
        ),
        a_log=torch.tensor([0.0, 0.5], dtype=torch.float32),
        dt_bias=torch.tensor([0.25, -0.5, 0.75, -0.25], dtype=torch.float32),
        b_proj=torch.tensor(
            [[0.5, -0.25, 0.25, 0.0], [0.0, 0.5, -0.5, 0.25]],
            dtype=torch.bfloat16,
        ),
        g_proj=torch.tensor(
            [
                [0.5, 0.0, 0.0, -0.25],
                [0.0, 0.5, 0.25, 0.0],
                [-0.25, 0.0, 0.5, 0.0],
                [0.0, 0.25, 0.0, 0.75],
            ],
            dtype=torch.bfloat16,
        ),
        o_norm=torch.tensor([1.0, 1.5], dtype=torch.float32),
        o_proj=identity,
    )


def _tokens() -> torch.Tensor:
    return torch.tensor(
        [[[0.5, -1.0, 0.25, 0.75], [-0.25, 0.5, 1.0, -0.5]]],
        dtype=torch.bfloat16,
    )


def test_official_kda_full_and_incremental_paths_match_all_final_state() -> None:
    cfg = _config()
    weights = _weights()
    tokens = _tokens()
    zero = zero_official_kda_state(cfg, batch_size=1, device=tokens.device)

    full = official_kda(tokens, weights, zero, cfg)
    first = official_kda(tokens[:, :1], weights, zero, cfg)
    second = official_kda(tokens[:, 1:], weights, first.state, cfg)

    torch.testing.assert_close(
        full.output,
        torch.cat((first.output, second.output), dim=1),
        atol=0,
        rtol=0,
    )
    torch.testing.assert_close(full.state.conv_q, second.state.conv_q, atol=0, rtol=0)
    torch.testing.assert_close(full.state.conv_k, second.state.conv_k, atol=0, rtol=0)
    torch.testing.assert_close(full.state.conv_v, second.state.conv_v, atol=0, rtol=0)
    torch.testing.assert_close(
        full.state.recurrent_v_first,
        second.state.recurrent_v_first,
        atol=1.0e-6,
        rtol=1.0e-6,
    )
    assert full.output.dtype == torch.bfloat16
    assert full.state.conv_q.dtype == torch.bfloat16
    assert full.state.recurrent_v_first.dtype == torch.float32
    assert full.boundaries.q.shape == (1, 2, 2, 2)
    assert full.boundaries.log_decay.shape == (1, 2, 2, 2)
    assert full.boundaries.beta.shape == (1, 2, 2)
    assert all(
        torch.isfinite(value).all()
        for value in full.boundaries.__dict__.values()
    )


def test_official_kda_recurrence_decays_key_axis_before_delta_update() -> None:
    cfg = _config()
    weights = _weights()
    token = _tokens()[:, :1]
    state = OfficialKdaState(
        conv_q=torch.tensor(
            [[[0.25, -0.5, 0.75, -1.0], [0.5, 0.25, -0.25, 0.75]]],
            dtype=torch.bfloat16,
        ),
        conv_k=torch.tensor(
            [[[-0.5, 0.25, 0.5, -0.75], [0.25, -0.5, 0.75, 0.5]]],
            dtype=torch.bfloat16,
        ),
        conv_v=torch.tensor(
            [[[0.75, -0.25, 0.5, 0.25], [-0.5, 0.75, -0.25, 0.5]]],
            dtype=torch.bfloat16,
        ),
        recurrent_v_first=(
            torch.arange(8, dtype=torch.float32).reshape(1, 2, 2, 2) / 32.0
        ),
    )
    original = tuple(value.clone() for value in state.__dict__.values())

    result = official_kda(token, weights, state, cfg)

    q = result.boundaries.q[:, 0].float()
    k = result.boundaries.k[:, 0].float()
    v = result.boundaries.v[:, 0].float()
    alpha = result.boundaries.log_decay[:, 0].float().exp()
    beta = result.boundaries.beta[:, 0].float()
    state_kv = state.recurrent_v_first.transpose(-1, -2)
    decayed = alpha.unsqueeze(-1) * state_kv
    prediction = torch.einsum("bhk,bhkv->bhv", k, decayed)
    delta = (v - prediction) * beta.unsqueeze(-1)
    expected_kv = decayed + k.unsqueeze(-1) * delta.unsqueeze(-2)
    expected_output = torch.einsum("bhk,bhkv->bhv", q, expected_kv)

    torch.testing.assert_close(
        result.state.recurrent_v_first,
        expected_kv.transpose(-1, -2),
        atol=1.0e-6,
        rtol=1.0e-6,
    )
    torch.testing.assert_close(
        result.boundaries.recurrent_output[:, 0],
        expected_output,
        atol=1.0e-6,
        rtol=1.0e-6,
    )
    for before, after in zip(original, state.__dict__.values()):
        torch.testing.assert_close(before, after, atol=0, rtol=0)


@pytest.mark.parametrize(
    "weights",
    (
        replace(_weights(), a_log=torch.zeros(2, dtype=torch.bfloat16)),
        replace(_weights(), a_log=torch.zeros(3, dtype=torch.float32)),
        replace(_weights(), q_proj=torch.eye(3, dtype=torch.bfloat16)),
        replace(_weights(), q_proj=object()),
    ),
)
def test_official_kda_rejects_dtype_or_shape_drift(weights: OfficialKdaWeights) -> None:
    cfg = _config()
    tokens = _tokens()
    state = zero_official_kda_state(cfg, batch_size=1, device=tokens.device)

    with pytest.raises(ValueError, match="invalid official KDA"):
        official_kda(tokens, weights, state, cfg)


def test_official_kda_rejects_k_first_labeled_state_shape() -> None:
    cfg = _config()
    tokens = _tokens()
    zero = zero_official_kda_state(cfg, batch_size=1, device=tokens.device)
    bad = replace(zero, recurrent_v_first=torch.zeros((1, 2, 3, 2)))

    with pytest.raises(ValueError, match="invalid official KDA"):
        official_kda(tokens, _weights(), bad, cfg)


def test_official_kda_rejects_wrong_convolution_history_width() -> None:
    cfg = _config()
    tokens = _tokens()
    zero = zero_official_kda_state(cfg, batch_size=1, device=tokens.device)
    bad = replace(zero, conv_q=torch.zeros((1, 3, 4), dtype=torch.bfloat16))

    with pytest.raises(ValueError, match="invalid official KDA"):
        official_kda(tokens, _weights(), bad, cfg)


def test_official_kda_rejects_non_finite_weight() -> None:
    cfg = _config()
    tokens = _tokens()
    zero = zero_official_kda_state(cfg, batch_size=1, device=tokens.device)
    bad = replace(_weights(), dt_bias=torch.tensor([0.0, float("nan"), 0.0, 0.0]))

    with pytest.raises(ValueError, match="invalid official KDA"):
        official_kda(tokens, bad, zero, cfg)


def test_official_kda_rejects_empty_sequence() -> None:
    cfg = _config()
    tokens = _tokens()[:, :0]
    zero = zero_official_kda_state(cfg, batch_size=1, device=tokens.device)

    with pytest.raises(ValueError, match="invalid official KDA"):
        official_kda(tokens, _weights(), zero, cfg)
def test_official_kda_projection_is_sequence_partition_invariant() -> None:
    torch.manual_seed(20260811)
    hidden = torch.randn(1, 2, 2_048, dtype=torch.float32).to(torch.bfloat16)
    weight = torch.randn(1_024, 2_048, dtype=torch.float32).to(torch.bfloat16)

    full = _project(hidden, weight)
    incremental = torch.cat(
        tuple(_project(hidden[:, index : index + 1], weight) for index in range(2)),
        dim=1,
    )

    assert torch.equal(full, incremental)


def test_official_kda_accepts_matvec_projection_weights() -> None:
    class MatvecWeight:
        def __init__(self, tensor: torch.Tensor) -> None:
            self.tensor = tensor
            self.shape = tensor.shape
            self.dtype = tensor.dtype
            self.device = tensor.device

        def matvec(self, value: torch.Tensor) -> torch.Tensor:
            return torch.nn.functional.linear(
                value.reshape(1, -1), self.tensor
            ).reshape(-1)

    cfg = _config()
    tokens = _tokens()
    weights = _weights()
    zero = zero_official_kda_state(cfg, batch_size=1, device=tokens.device)
    expected = official_kda(tokens, weights, zero, cfg)
    actual = official_kda(
        tokens,
        replace(weights, q_proj=MatvecWeight(weights.q_proj)),
        zero,
        cfg,
    )

    assert torch.equal(actual.output, expected.output)
    assert torch.equal(actual.state.conv_q, expected.state.conv_q)
    assert torch.equal(
        actual.state.recurrent_v_first, expected.state.recurrent_v_first
    )
